### PR TITLE
Superfluid And AddressBook Ceramic Datamodel Integration

### PR DESCRIPTION
**Pull request by carlbarrdahl**
https://www.loom.com/share/2638383e2407493bbead8c980e5e5f1d

Right now the Connect to Ceramic button is displayed in the AddressBook and SearchDialog components. In the future this could be moved to an Avatar in the SfAppBar component.

I tried a few different approaches in syncing the data to Ceramic. The cleanest one would be to listen to changes to the addressBookEntries redux state and simply merge with the Ceramic state. This caused some strange behaviours because it tried sync more often than necessary.

Next I tried to manually trigger a sync method from the AddressBook component. I figured I would be able to just use the redux state (addressBookEntries) but the updated state wasn't available when the function was triggered.

Finally I used two methods (add and remove) and either concatenated or filtered to get the complete state.

Another approach could be to hook into redux-persist middleware but I couldn't figure out how to access the Ceramic internals (Ceramic or SelfID) without using hooks.

It usually takes a few seconds (on my network) to sync the data to Ceramic and to improve the UX there could be a loading spinner indicating a sync in progress. This could be placed either next to the copy or star icon in the AccountsPage (`pages/[_network]/accounts/[_id].tsx`) or in SfAppBar top right.

Link to Gitcoin: https://gitcoin.co/issue/28896
Demo website: https://superfluid-console-liart.vercel.app/

### CODE REVIEW

**Comment by kasparkallas** [comment]
Hey @carlbarrdahl

Good job so far, it looks cool and promising!

But I have a burning question, is this solution impersonation proof? I don't see any signing anywhere and just connecting a wallet is not a safe solution for account authentication (e.g. see https://www.impersonator.xyz/).

If it is then can you please describe how it works.

**Comment by carlbarrdahl** [comment]
> Hey @carlbarrdahl
> 
> Appreciate the effort so far -- looks cool and promising!
> 
> But I have a burning question, is this solution impersonation proof? I don't see any signing anywhere and just connecting a wallet is not a safe solution for account authentication (e.g. see https://www.impersonator.xyz/).
> 
> If it is then can you please describe how it works.

Thank you!
Valid question. It's using Ceramic with Self.ID (which uses DIDs: https://github.com/ceramicstudio/js-3id).

The reason you didn't see a wallet sign in the demo was because I had previously signed in Metamask for this wallet.

You can test how this works at: https://self.id

**Comment by kasparkallas** [comment]
> Thank you @kasparkallas! Valid question. It's using Ceramic with Self.ID (which uses DIDs: https://github.com/ceramicstudio/js-3id).
> 
> The reason you didn't see a wallet sign in the demo was because I had previously signed in Metamask for this wallet.
> 
> You can test how this works at: https://self.id

That's great to hear! @carlbarrdahl After the authentication, do you know whether it uses a cookie or a JWT/bearer token to persist the session? Would appreciate a demo with the full flow.

Regarding the syncing of the data, you mentioned merging, is it also possible to have more operational granularity with Ceramic (i.e. instead of merging the whole state, do specific calls such as "add entry", "edit entry" etc. )?

**Comment by kasparkallas** [comment]
@carlbarrdahl This branch is now also deployed to [superfluid-console-gu92l5u5l-superfluid-finance.vercel.app](https://superfluid-console-gu92l5u5l-superfluid-finance.vercel.app/) now. https://superfluid-console-liart.vercel.app/ is a bit broken because we have CORS limitations set for the Superfluid RPC-s.

**Comment by vmichalik** [comment]
Hey, awesome work - just trying it out. Doesn't seem like you can restart the connect to Ceramic flow currently if you fail out of it (e.g. by not signing in metamask), as the connect button disappears.

**Comment by carlbarrdahl** [comment]
Thank you for the great questions and feedback @kasparkallas and @vmichalik!

> Regarding the syncing of the data, you mentioned merging, is it also possible to have more operational granularity with Ceramic (i.e. instead of merging the whole state, do specific calls such as "add entry", "edit entry" etc.)?

As far as I'm concerned there are 3 methods available in the api - get, set and merge (only shallow) (https://developers.ceramic.network/reference/self-id/classes/web.SelfID/#set). 


I agree that it would be great with more granular control over data. Perhaps @oed or @PaulLeCam have more information on this? 

> do you know whether it uses a cookie or a JWT/bearer token to persist the session? Would appreciate a demo with the full flow.

I believe it's using cookies to persist the session (https://github.com/ceramicstudio/self.id/blob/main/packages/react/src/storage.ts). Again, the Ceramic team might be able to provide more information on this.

Will see if I can record a new video tomorrow.

Another thing I've noticed is it's not totally clear on how to unlink an account or app. For example to reset the state I have to use a new wallet address which makes testing and debugging harder. 

> Hey, awesome work - just trying it out. Doesn't seem like you can restart the connect to Ceramic flow currently if you fail out of it (e.g. by not signing in metamask), as the connect button disappears.

Fixed the issue with disappearing connect button and now displays an Alert error message.

Thanks for pushing the deployment to fix CORS issue @kasparkallas!

File: src/components/ConnectCeramicButton.tsx [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    console.log("Error connecting wallet", error);
+  }
+}
+
+// Sync state with Ceramic.
+// Another approach would be to listen to changes in addressBookEntries but
+// wasn't able to make work in a good way.
+export function useAdressBookSync() {
+  // @ts-ignore
+  const record = useViewerRecord("myAddressBook");
```
Comment (kasparkallas) [inline src/components/ConnectCeramicButton.tsx:48]: The data is encrypted and secured on Ceramic's end so the "viewer record" is only visible to the authenticated identity, right?

I would appreciate a link where this is stated explicitly as well :pray: [carlbarrdahl](https://github.com/carlbarrdahl)

**Comment by vmichalik** [comment]
We don't currently plan to add this feature

**Pull request closed without merging**